daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -1267,25 +1213,69 @@ private[lf] object SBuiltin {
     }
   }
 
-  /** $fetchKey[T]
-    *   :: { key: key, maintainers: List Party }
-    *   -> ContractId T
-    */
-  final case class SBUFetchKey(templateId: TypeConName) extends OnLedgerBuiltin(1) {
+  private[speedy] sealed abstract class SBUKeyBuiltin(templateId: TypeConName)
+      extends OnLedgerBuiltin(1) {
     private[this] val typ = AstUtil.TContractId(Ast.TTyCon(templateId))
-    override protected final def execute(
+
+    sealed trait KeyOperation {
+      // Callback from the engine returned NotFound
+      def handleKeyNotFound(machine: Machine): Boolean
+      // We saw an archive of this key.
+      def handleKeyArchived(machine: Machine, key: GlobalKey): Unit
+      def cidToSValue(cid: V.ContractId): SValue
+      def cidToSExpr(cid: V.ContractId): SExpr
+    }
+    final case object Fetch extends KeyOperation {
+      override def handleKeyNotFound(machine: Machine): Boolean =
+        machine.tryHandleSubmitMustFail()
+      override def handleKeyArchived(machine: Machine, key: GlobalKey): Unit =
+        // TODO (MK) Produce a proper error here.
+        crash(s"Could not find key $key")
+      override def cidToSValue(cid: V.ContractId) =
+        SContractId(cid)
+      override def cidToSExpr(cid: V.ContractId) =
+        SEImportValue(typ, V.ValueContractId(cid))
+    }
+    final case object Lookup extends KeyOperation {
+      override def handleKeyNotFound(machine: Machine): Boolean = {
+        machine.returnValue = SV.None
+        true
+      }
+      override def handleKeyArchived(machine: Machine, key: GlobalKey): Unit = {
+        machine.returnValue = SV.None
+      }
+      override def cidToSValue(cid: V.ContractId) =
+        SOptional(Some(SContractId(cid)))
+      override def cidToSExpr(cid: V.ContractId) =
+        SBSome(SEImportValue(typ, V.ValueContractId(cid)))
+    }
+
+    protected final def lookupKey(
```

Comment:
suggest inline `lookupKey` and `fetchKey` into their sole respective callers, which can call `handleKey` directly.
(`handleKey` will need to become `protected final` instead of `private`.)